clippy_lints/src/utils/higher.rs

```diff
@@ -114,13 +114,36 @@ pub fn range(expr: &hir::Expr) -> Option<Range> {
 
 /// Checks if a `let` decl is from a `for` loop desugaring.
 pub fn is_from_for_desugar(decl: &hir::Decl) -> bool {
+    // This will detect plain for-loops without an actual variable binding:
+    //
+    // ```
+    // for x in some_vec {
+    //   // do stuff
+    // }
+    // ```
     if_let_chain! {[
         let hir::DeclLocal(ref loc) = decl.node,
         let Some(ref expr) = loc.init,
         let hir::ExprMatch(_, _, hir::MatchSource::ForLoopDesugar) = expr.node,
     ], {
         return true;
     }}
+
+    // This detects a variable binding in for loop to avoid `let_unit_value`
```

Comment:
I think that's a different `DeclLocal` than the one we're looking for, but it should probably also contain a `source` other than `Normal`. I'll see if I can produce a PR for rustc.

I think it'll call `is_from_for_desugar` if you have a `let a = 42;` outside the for loop, which will trigger the shadowing lints.

This PR is fine as it is, since this is a non-issue for the `let_unit_value` lint anyway